### PR TITLE
Doc WinAppSDK's WinRT registration choice

### PR DESCRIPTION
**Pull request by DrusTheAxe**
This is a polished form of the information in issue #2352 [WinAppSdk is difficult to use with reg-free WinRT due to class/file naming, pairs](https://github.com/microsoft/WindowsAppSDK/issues/2352#)
 and particularly [in this comment](https://github.com/microsoft/WindowsAppSDK/issues/2352#issuecomment-1087004638).

### CODE REVIEW

**Review by dhoehna — APPROVED** [review]

File: docs/Coding-Guidelines/WinRT-Registration.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+* rename the libraries, creating a saner layout, adding numbers to "dupes" (e.g. UI.Xaml.1, UI.Xaml.2, ...)
+* adjust activation factory search algorithm as needed
+
+This is a projection-specific answer, requires new designs and complicated new machinery, and
+doesn't solve all the problems without devolving to #1 but with more work.
+
+These options were discussed at length (in 2021) before deciding to embrace manifested solutions
+instead of language projections' manifest-free option. Not all component authors face the same
+choices so YMMV but, for WinAppSDK, this was (and continues to be) an intentional decision for good
+reasons
```
[the remark this answers is not in the document]
Comment (BenJKuhn) [reply docs/Coding-Guidelines/WinRT-Registration.md:144]: Pointer? I'd love to read what you've written, @riverar. Registration models are a hotly debated topic around here and I appreciate seeing different opinions.

I can't speak to whether Howard was inspired by what you wrote, but I can confirm that we've discussed all of these (and more) at various points. We've also talked about augmenting metadata (it is designed for fast lookup anyway) and using alternate data registries. Most were ideation though and didn't get any real traction.

**Review by BenJKuhn — APPROVED** [review]

File: docs/Coding-Guidelines/WinRT-Registration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (riverar) [reply docs/Coding-Guidelines/WinRT-Registration.md:144]: It's a cut/paste of what I wrote in my issue https://github.com/microsoft/WindowsAppSDK/issues/2352#:~:text=Considered%20workarounds/solutions

[the remark this answers is not in the document]
Comment (riverar) [reply docs/Coding-Guidelines/WinRT-Registration.md:144]: I'm not claiming any of this is novel or any sort of ownership. My only concern is that this section proports to be a record of past considerations but then uses text from yesterday. If you're saying you discussed these _exact_ solutions in the past then I'm OK with it!

[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply docs/Coding-Guidelines/WinRT-Registration.md:144]: > If you're saying you discussed these _exact_ solutions

Yup!

And I'd swear the was at least one other solution but I haven't been able to put my finger on it (yet ;-)

The options were discussed, and at some length, over an extended period of time, and maybe not a cast of thousands &lt;g&gt; but involving multiple groups and times. Unfortunately the options, tradeoffs, discussion and ultimately the decision wasn't _documented_ :-(

As you might guess, you're not the first to raise these questions. But this did give us an opportunity to fill a doc gap we should have already plugged. Better late than never. So for that, consider this a warm thank you :-)

**Review by raghavthind2005 — APPROVED** [review]

File: docs/Coding-Guidelines/WinRT-Registration.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+caller uses a language project (C#/WinRT, C++/WinRT, etc), the WinRT runtimeclass implementation is
+in a DLL and the DLL's filename matches the ACID's namespace/prefix.
+
+WinAppSDK/MSIX uses option A.
+
+WinAppSDK/Self-Contained uses option B.
+
+## Why use manifested registration?
+
+WinAppSDK uses options A+B ByDesign, for performance and complexity reasons.
```
Comment (jonwis) [inline docs/Coding-Guidelines/WinRT-Registration.md:32]: Just state the two names here, "WinAppSDK uses both the process' package graph and registration-free WinRT." Then go on to explain the precise reasoning if you see fit. No need to tease the "for perf and cplx" parts.

File: docs/Coding-Guidelines/WinRT-Registration.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+While some might call this "non-ideal layout" (implying projections' manifest-free resolution is
+ideal), that's not true. The manifest-free option is *simple*, but not necessarily best. As in
+WinAppSDK's case.
```
[the remark this answers is not in the document]
Comment (jonwis) [reply docs/Coding-Guidelines/WinRT-Registration.md:36]: Just delete and assert what we do and why. No need to defend it.  If you are so motivated, you could write an appendix on how projection-based lookup doesn't meet our current needs w.r.t. naming of components.

File: docs/Coding-Guidelines/WinRT-Registration.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+WinAppSDK policy and practices doesn't doesn't match
```
Comment (jonwis) [inline docs/Coding-Guidelines/WinRT-Registration.md:38]: "But why don't you just fix that so you can use it?" - consider replacing this with a simpler statement:

> Manifest-based registration allows explicit binding of type names to implementations.  Windows App SDK's mapping of types to implementation DLLs may change over time, and being explicit ensures the mapping logic is centralized.

**Review by jonwis — COMMENTED** [review]
This doc reads like a justification of the choice made, not a documentation.  "We needed to accomplish X, our options were Q and Y and Z, option Z met the criteria for X."

**Comment by riverar** [comment]
@DrusTheAxe I can open a separate issue on this if needed, but the embedded fusion manifest requirement appears to [deviate from documented expectations](https://docs.microsoft.com/cpp/build/concepts-of-isolated-applications-and-side-by-side-assemblies?view=msvc-170#manifests-and-search-order), in that the manifest can be embedded as a resource or be on disk, as a fallback. The embedded-only approach that WAS has implemented is a pain for Rust that has no native tooling to embed a manifest.

Was this just deemed unnecessary at the time?

File: docs/Coding-Guidelines/WinRT-Registration.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
+# WinRT Registration
+
+Windows App SDK provides registers WinRT types via *manifested* solutions:
```
[the remark this answers is not in the document]
Comment (lyahdav) [reply docs/Coding-Guidelines/WinRT-Registration.md:3]: Looks like this comment wasn't addressed

[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply docs/Coding-Guidelines/WinRT-Registration.md:3]: Good catch! Thanks

Fixed in https://github.com/microsoft/WindowsAppSDK/pull/3094

**Pull request merged**